fix: Use dequeue to track WindowPartitions in RowStreamingWindowBuild

JkSelf opened the pull request:
In the core case, if each row corresponds to a separate WindowPartition, this could result in an excessive number of WindowPartition instances being held in memory, potentially leading to OOM . To fix this issue, this PR  proposes replacing the std::vector with a deque data structure, which will allow for more efficient release of WindowPartition resources.

aditi-pandit commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -102,6 +102,14 @@ void RowsStreamingWindowBuild::noMoreInput() {
 
 std::shared_ptr<WindowPartition> RowsStreamingWindowBuild::nextPartition() {
   VELOX_CHECK(hasNextPartition());
+
```
@JkSelf : So without this change you were accumulating all the partitions, when it should've been streaming behavior.

But doesn't seem like those partitions are holding onto any substantial memory, else we would've caught this before.

Please can you clarify the reason for this fix.

FelixYBW commented:
In Velox we track the memory by plan node memory pool or global spill memory pool. In gluten the memory pool for plan node is counted into offheap size, the global spill memory pool is counted into overhead. But there are much memory allocation in Velox which isn't tracked. The common issue is to allocate a potential per row std::vector which can be very large. Because each partition can have like 1G row, then even a std::vector<rowVector*> can take 8G memory.

The solution is 1) try best to avoid the per row std::vector 2) if we have to use the per row std::vector, define it as std::vector<X, memory::StlAllocator<char*>>.

To this issue, it's common case that each partition has one row, so eventually it's a per row std::vector.

aditi-pandit commented:
> In Velox we track the memory by plan node memory pool or global spill memory pool. In gluten the memory pool for plan node is counted into offheap size, the global spill memory pool is counted into overhead. But there are much memory allocation in Velox which isn't tracked. The common issue is to allocate a potential per row std::vector which can be very large. Because each partition can have like 1G row, then even a std::vector<rowVector*> can take 8G memory.
> 
> The solution is 1) try best to avoid the per row std::vector 2) if we have to use the per row std::vector, define it as std::vector<X, memory::StlAllocator<char*>>.
> 
> To this issue, it's common case that each partition has one row, so eventually it's a per row std::vector.

Thanks @FelixYBW. So you are saying the memory consumption comes from https://github.com/facebookincubator/velox/blob/main/velox/exec/WindowPartition.h#L239.

The fix is fine imo.

But maybe we should also define rows_ with StlAllocator.

JkSelf commented:
> > In Velox we track the memory by plan node memory pool or global spill memory pool. In gluten the memory pool for plan node is counted into offheap size, the global spill memory pool is counted into overhead. But there are much memory allocation in Velox which isn't tracked. The common issue is to allocate a potential per row std::vector which can be very large. Because each partition can have like 1G row, then even a std::vector<rowVector*> can take 8G memory.
> > The solution is 1) try best to avoid the per row std::vector 2) if we have to use the per row std::vector, define it as std::vector<X, memory::StlAllocator<char*>>.
> > To this issue, it's common case that each partition has one row, so eventually it's a per row std::vector.
> 
> Thanks @FelixYBW. So you are saying the memory consumption comes from https://github.com/facebookincubator/velox/blob/main/velox/exec/WindowPartition.h#L239.
> 
> The fix is fine imo.
> 
> But maybe we should also define rows_ with StlAllocator.

@aditi-pandit I will file another pr to add the StlAllocator after this pr merged. Thanks.

aditi-pandit reviewed (approved):
Thanks @JkSelf

JkSelf commented:
@xiaoxmeng Can you help to merge this PR? Thanks.

FelixYBW commented:
> @xiaoxmeng Can you help to merge this PR? Thanks.

Hold on please. I haven't verified.

aditi-pandit commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -102,6 +102,15 @@ void RowsStreamingWindowBuild::noMoreInput() {
 std::shared_ptr<WindowPartition> RowsStreamingWindowBuild::nextPartition() {
   VELOX_CHECK(hasNextPartition());
+
+  if (outputPartition_ > 0) {
```
This condition is there because you don't want to erase the first partition rightaway.

As I think more about it, I would rewrite this code so that if windowPartitions_ .size() > 2, then you now that you have a partition that has been output already. So then you clean it.

FelixYBW replied:
I'd think we should use list or map. vector isn't good here. Or more efficient a ring vector to reuse the entry, instead of erase it.

FelixYBW commented:
@aditi-pandit can you take a look again? It's an important fix for common OOM issue from customer side.

aditi-pandit commented:
@JkSelf : Can we have a test for this code specifically ? Would be great if you can add something.

JkSelf commented:
> @JkSelf : Can we have a test for this code specifically ? Would be great if you can add something.

@aditi-pandit The existing unit tests [here ](https://github.com/facebookincubator/velox/blob/main/velox/exec/tests/WindowTest.cpp#L128-L295)can verify this changes.

aditi-pandit commented:
> > @JkSelf : Can we have a test for this code specifically ? Would be great if you can add something.
> 
> @aditi-pandit The existing unit tests [here ](https://github.com/facebookincubator/velox/blob/main/velox/exec/tests/WindowTest.cpp#L128-L295)can verify this changes.

@JkSelf : Agree. But I would've expected this error to be caught by a test like https://github.com/facebookincubator/velox/blob/main/velox/exec/tests/WindowTest.cpp#L154

It would be great to write a test with the previously failing input data so that we are assured that this code is working.

FelixYBW commented on velox/exec/RowsStreamingWindowBuild.h:
```diff
@@ -73,13 +72,13 @@ class RowsStreamingWindowBuild : public WindowBuild {
 
   // Holds all the built window partitions.
-  std::vector<std::shared_ptr<WindowPartition>> windowPartitions_;
+  std::deque<std::shared_ptr<WindowPartition>> windowPartitions_;
```
with next fix, deque header is the input partition which may be empty or already complete, tail is the output partition which may be output already or going to be output next.

FelixYBW commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -41,26 +41,37 @@ RowsStreamingWindowBuild::RowsStreamingWindowBuild(
 
   if (finished) {
-    windowPartitions_[inputPartition_]->setComplete();
+    windowPartitions_.back()->setComplete();
```
Add code before ++inputPartition_:
```
    if (outputPartition_ == inputPartition_){
      //if the output partition is the input partition, it's already output so we need pop the output partition.
      windowPartitions_.pop_front();
    }
```

JkSelf replied:
@FelixYBW Yes. Resolved.

FelixYBW commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -102,12 +113,15 @@ void RowsStreamingWindowBuild::noMoreInput() {
-  return windowPartitions_[++outputPartition_];
+  outputPartition_++;
+  auto output = std::move(windowPartitions_.front());
+  windowPartitions_.pop_front();
```
change to:
```
  // it's possible the output partition is the input partition
  // and input partition isn't completed, so we don't pop
  if (outputPartition_ < inputPartition_){
    windowPartitions_.pop_front();
  }
```

JkSelf replied:
@FelixYBW Added.

JkSelf commented on velox/exec/RowsStreamingWindowBuild.h:
```diff
@@ -73,13 +72,13 @@ class RowsStreamingWindowBuild : public WindowBuild {
 
   // Point to the current output partition if not -1.
-  vector_size_t outputPartition_ = -1;
+  vector_size_t outputPartition_ = 0;
```
@aditi-pandit Yes. Updated.

JkSelf commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -41,26 +41,29 @@ RowsStreamingWindowBuild::RowsStreamingWindowBuild(
+    windowPartitions_.back()->setComplete();
     ++inputPartition_;
+    // Create a new partition for the next input.
+    windowPartitions_.emplace_back(std::make_shared<WindowPartition>(
```
@aditi-pandit Yes. Updated.

JkSelf commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -41,26 +41,29 @@ RowsStreamingWindowBuild::RowsStreamingWindowBuild(
-  }
-
-  windowPartitions_[inputPartition_]->addRows(inputRows_);
+  windowPartitions_.back()->addRows(inputRows_);
```
@aditi-pandit Yes. Added.

JkSelf commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -102,12 +105,14 @@ void RowsStreamingWindowBuild::noMoreInput() {
 std::shared_ptr<WindowPartition> RowsStreamingWindowBuild::nextPartition() {
   VELOX_CHECK(hasNextPartition());
-  return windowPartitions_[++outputPartition_];
+  outputPartition_++;
```
@aditi-pandit Yes. Updated.

JkSelf commented:
@aditi-pandit @FelixYBW Resolved all your comments. Can you help to review again? Thanks. I updated the pool capacity from 8MB to 4MB [here ](https://github.com/facebookincubator/velox/pull/11077/files#diff-9f41a5e1a8da742fc53f54ba85588678cdb81ffb05b38eefebe21ad5138def8eR183)and the oom will occur without this PR.

FelixYBW commented:
> @aditi-pandit @FelixYBW Resolved all your comments. Can you help to review again? Thanks. I updated the pool capacity from 8MB to 4MB [here ](https://github.com/facebookincubator/velox/pull/11077/files#diff-9f41a5e1a8da742fc53f54ba85588678cdb81ffb05b38eefebe21ad5138def8eR183)and the oom will occur without this PR.

Thank you! Ke. Did you also add test to have a large window partition? It's used to test that input and output partition are the same one.

JkSelf commented:
> > @aditi-pandit @FelixYBW Resolved all your comments. Can you help to review again? Thanks. I updated the pool capacity from 8MB to 4MB [here ](https://github.com/facebookincubator/velox/pull/11077/files#diff-9f41a5e1a8da742fc53f54ba85588678cdb81ffb05b38eefebe21ad5138def8eR183)and the oom will occur without this PR.
> 
> Thank you! Ke. Did you also add test to have a large window partition? It's used to test that input and output partition are the same one.

@FelixYBW  I added the test in https://github.com/facebookincubator/velox/pull/11077/files#diff-9f41a5e1a8da742fc53f54ba85588678cdb81ffb05b38eefebe21ad5138def8eR221 to evaluate huge window partitions. Based on previous experience https://github.com/facebookincubator/velox/pull/10883, this test may cause issues with the meta CI when dealing with large window partitions.
My suggestion is to use the existing rowBasedStreamingWindowOOM test and reduce the memory capacity from 8MB to 4MB.

aditi-pandit commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -102,12 +123,22 @@ void RowsStreamingWindowBuild::noMoreInput() {
+  // it means the current output partition has been processed and there are new
+  // input partitions. Therefore, we need to remove the processed output
+  // partition from the queue.
+  if (outputPartition_ < inputPartition_) {
```
This is a bit counter-intuitive to understand... This code should generally pop the windowPartitions_ except when outputPartition_ == inputPartition_.

Do we ever encounter the case when outputPartition_ > inputPartition_ ?

JkSelf replied:
@aditi-pandit  The outputPartition_ is always less than or equal to inputPartition_ and will never exceed it. It seems we can remove the code in addInput that handles the removal of processed partitions and keep this logic only in nextPartition(). cc @FelixYBW

FelixYBW replied:
If we remove the pop from addPartitionInputs, the outputPartition_ may be output twice, will it lead to error?

```
    if (outputPartition_ == inputPartition_) {
      // If the output partition is the same as the input partition, it has
      // already been output,
      // so we need to remove the output partition from the queue.
      windowPartitions_.pop_front();
    }
```

aditi-pandit commented on velox/exec/tests/WindowTest.cpp:
```diff
@@ -217,6 +217,64 @@ TEST_F(WindowTest, rowBasedStreamingWindowOOM) {
 }
 
+TEST_F(WindowTest, windowOOMWithHugeWindowPartition) {
+  const vector_size_t size = 1'000'000'00;
```
Could we work with smaller sizes ?

JkSelf commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -41,26 +41,47 @@ RowsStreamingWindowBuild::RowsStreamingWindowBuild(
+      (windowPartitions_.front()->numRows() == 0);
+}
+
+std::shared_ptr<WindowPartition> RowsStreamingWindowBuild::inputPartition() {
```
@aditi-pandit Yes. Good suggestions. Updated.

FelixYBW commented:
@aditi-pandit Any more comments?

aditi-pandit commented on velox/exec/tests/WindowTest.cpp:
```diff
@@ -217,6 +217,64 @@ TEST_F(WindowTest, rowBasedStreamingWindowOOM) {
+    }
+  };
+  // RowStreamingWindow will not OOM even with huge window partition.
+  testWindowBuild(true);
```
Did you intend to call testWindowBuild(false) as well ?

aditi-pandit reviewed (commented):
Thanks @JkSelf.

Just minor comment left.

FelixYBW commented:
The current PR can't fix the window OOM issue any more. Let me check.

liujiayi771 commented:
> The current PR can't fix the window OOM issue any more. Let me check.

I have a test case where running on window results in an OOM error. After applying this PR, the OOM issue no longer occurs.

aditi-pandit commented on velox/exec/tests/WindowTest.cpp:
```diff
@@ -160,14 +160,28 @@ TEST_F(WindowTest, spillUnsupported) {
+          makeFlatVector<int64_t>(size, [](auto row) { return row; }),
+           // Payload.
+           makeFlatVector<int64_t>(size, [](auto row) { return row; }),
+           // Partition key.
```
Don't follow the use of // Partition key and //Sorting key for these columns...Are they used in PARTITION BY and ORDER BY clauses anywhere ?

JkSelf replied:
@aditi-pandit Just for local test. And I have removed this code.

aditi-pandit reviewed (commented):
Thanks @JkSelf. Have few comments.

JkSelf commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -102,12 +120,22 @@ void RowsStreamingWindowBuild::noMoreInput() {
-  return windowPartitions_[++outputPartition_];
+  auto output = outputPartition();
+  outputPartition_++;
+  // // Remove the processed output partition from the queue.
```
@aditi-pandit Sure. Removed.

JkSelf commented on velox/exec/RowsStreamingWindowBuild.h:
```diff
@@ -72,14 +73,17 @@ class RowsStreamingWindowBuild : public WindowBuild {
-  // Point to the current output partition if not -1.
-  vector_size_t outputPartition_ = -1;
+  // Point to the current output partition.
+  vector_size_t outputPartition_ = 0;
```
@aditi-pandit Sure. Removed  `outputPartition_ ` and `inputPartition_`.

JkSelf commented:
@xiaoxmeng @aditi-pandit @FelixYBW 
I attempted to increase the number of rows while reducing the memory capacity. However, this approach failed when creating the RowVector with lower memory capacity during unit testing. As a result, it is challenging to include an OOM case in the unit tests. We have submitted [PR](https://github.com/apache/incubator-gluten/pull/8757) in Gluten to test the unit tests, and @liujiayi771 has also helped verify that this PR can resolve the window OOM issue in their environment. Given these circumstances, could we proceed with merging this PR without the unit tests?

xiaoxmeng commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -102,12 +118,16 @@ void RowsStreamingWindowBuild::noMoreInput() {
 
 std::shared_ptr<WindowPartition> RowsStreamingWindowBuild::nextPartition() {
   VELOX_CHECK(hasNextPartition());
```
We can't have check here as hasNextPartition will update windowPartitions_

xiaoxmeng commented on velox/exec/RowsStreamingWindowBuild.h:
```diff
@@ -51,18 +51,19 @@ class RowsStreamingWindowBuild : public WindowBuild {
-    return windowPartitions_.empty() ||
-        outputPartition_ == windowPartitions_.size() - 1;
-  }
+  inline bool needsInput() override;
```
drop inline

xiaoxmeng commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -41,26 +41,42 @@ RowsStreamingWindowBuild::RowsStreamingWindowBuild(
       this);
+
+  // Create the first WindowPartition.
+  windowPartitions_.emplace_back(std::make_shared<WindowPartition>(
```
We can just create on-demand as before?

xiaoxmeng commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -41,26 +41,42 @@ RowsStreamingWindowBuild::RowsStreamingWindowBuild(
     return;
   }
 
-  if (windowPartitions_.size() <= inputPartition_) {
```
```
void ensureInputPartition() {
  if (windowPartitions_.empty() || windowPartitions_.back()->complete()) { 
windowPartitions_.push_back(std::make_shared<WindowPartition(data_.get(), inversedInputChannels_, sortKeyInfo_));
  }
}

ensureInputPartition();
inputPartition()->addRows(inputRows_);
```

xiaoxmeng commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -41,26 +41,42 @@ RowsStreamingWindowBuild::RowsStreamingWindowBuild(
-  windowPartitions_[inputPartition_]->addRows(inputRows_);
+  inputPartition()->addRows(inputRows_);
 
   if (finished) {
```
```
 if (finished) {
    inputPartition()->setComplete();
}
```

xiaoxmeng commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -41,26 +41,42 @@ RowsStreamingWindowBuild::RowsStreamingWindowBuild(
+}
+
+std::shared_ptr<WindowPartition> RowsStreamingWindowBuild::inputPartition() {
+  return windowPartitions_.back();
```
```
VELOX_CHECK(!windowPartitions_.empty());
VELOX_CHECK(!windowPartitions_.back()->complete());
return windowPartitions_.back().get();
```

xiaoxmeng commented on velox/exec/RowsStreamingWindowBuild.h:
```diff
@@ -51,18 +51,19 @@ class RowsStreamingWindowBuild : public WindowBuild {
   void addPartitionInputs(bool finished);
 
+  // Returns the current input partition.
+  std::shared_ptr<WindowPartition> inputPartition();
```
```
WindowPartition* inputPartition() const;
WindowPartition* outputPartition() const;
```

JkSelf replied:
After modifying the function to WindowPartition* outputPartition() const, we encountered the following exception. This issue arises because both [Window#callResetPartition](https://github.com/facebookincubator/velox/blob/main/velox/exec/Window.cpp#L297) and [RowsStreamingWindowBuild#hasNextPartition](https://github.com/facebookincubator/velox/pull/11077/files#diff-5c6755d0934c31d00ce5698db29491de6e92094f1489e627ed61c376ffd68832R133) attempt to release the same Window Partition twice.

```
double free or corruption (out)
*** Aborted at 1740063354 (Unix time, try 'date -d @1740063354') ***
*** Signal 6 (SIGABRT) (0x3bcbfe) received by PID 3918846 (pthread TID 0x7fba0de29700)                          (linux TID 3918975) (maybe from PID 3918846, UID 0) (code: -6), stack trace: ***
    @ 0000000013b0d831 folly::symbolizer::(anonymous namespace)::signalHandler(int, si                         ginfo_t*, void*)
                       /mnt/DP_disk3/jk/projects/gluten/ep/build-velox/build/velox_ep/                         deps-download/folly/folly/experimental/symbolizer/SignalHandler.cpp:453
    @ 000000000001441f (unknown)
    @ 000000000004300b gsignal
    @ 0000000000022858 abort
    @ 000000000008d265 (unknown)
    @ 00000000000952fb (unknown)
    @ 0000000000096f9f (unknown)
    @ 000000000e722f50 std::_Sp_counted_ptr<facebook::velox::exec::WindowPartition*, (                         __gnu_cxx::_Lock_policy)2>::_M_dispose()
                       /usr/include/c++/11/bits/shared_ptr_base.h:348
                       -> /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/RowsStre                         amingWindowBuild.cpp
    @ 000000000bbf0248 std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release()
                       /usr/include/c++/11/bits/shared_ptr_base.h:168
                       -> /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/connectors/hi                         ve/FileHandle.cpp
    @ 000000000bbed3c6 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_coun                         t()
                       /usr/include/c++/11/bits/shared_ptr_base.h:705
                       -> /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/connectors/hi                         ve/FileHandle.cpp
    @ 000000000e6aa5eb std::__shared_ptr<facebook::velox::exec::WindowPartition, (__gn                         u_cxx::_Lock_policy)2>::~__shared_ptr()
                       /usr/include/c++/11/bits/shared_ptr_base.h:1154
                       -> /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/Window.c                         pp
    @ 000000000e6b8701 std::__shared_ptr<facebook::velox::exec::WindowPartition, (__gn                         u_cxx::_Lock_policy)2>::operator=(std::__shared_ptr<facebook::velox::exec::WindowParti                         tion, (__gnu_cxx::_Lock_policy)2>&&)
                       /usr/include/c++/11/bits/shared_ptr_base.h:1250
                       -> /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/Window.c                         pp
    @ 000000000e6ace5f std::shared_ptr<facebook::velox::exec::WindowPartition>::operat                         or=(std::shared_ptr<facebook::velox::exec::WindowPartition>&&)
                       /usr/include/c++/11/bits/shared_ptr.h:385
                       -> /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/Window.c                         pp
    @ 000000000e6a75b2 facebook::velox::exec::Window::callResetPartition()
                       /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/Window.cpp:                         297
    @ 000000000e6a8840 facebook::velox::exec::Window::callApplyLoop(int, std::shared_p                         tr<facebook::velox::RowVector> const&)
                       /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/Window.cpp:                         658
    @ 000000000e6a8a99 facebook::velox::exec::Window::getOutput()
                       /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/Window.cpp:                         710
    @ 000000000e30484f facebook::velox::exec::Driver::runInternal(std::shared_ptr<face                         book::velox::exec::Driver>&, std::shared_ptr<facebook::velox::exec::BlockingState>&, s                         td::shared_ptr<facebook::velox::RowVector>&)::{lambda()#5}::operator()() const
                       /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/Driver.cpp:                         601
    @ 000000000e30b11b void facebook::velox::exec::Driver::withDeltaCpuWallTimer<faceb                         ook::velox::exec::Driver::runInternal(std::shared_ptr<facebook::velox::exec::Driver>&,                          std::shared_ptr<facebook::velox::exec::BlockingState>&, std::shared_ptr<facebook::vel                         ox::RowVector>&)::{lambda()#5}>(facebook::velox::exec::Operator*, facebook::velox::Cpu                         WallTiming facebook::velox::exec::OperatorStats::*, facebook::velox::exec::Driver::run                         Internal(std::shared_ptr<facebook::velox::exec::Driver>&, std::shared_ptr<facebook::ve                         lox::exec::BlockingState>&, std::shared_ptr<facebook::velox::RowVector>&)::{lambda()#5                         }&&)
                       /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/Driver.cpp:                         1057
    @ 000000000e306e55 facebook::velox::exec::Driver::runInternal(std::shared_ptr<face                         book::velox::exec::Driver>&, std::shared_ptr<facebook::velox::exec::BlockingState>&, s                         td::shared_ptr<facebook::velox::RowVector>&)
                       /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/Driver.cpp:                         598
    @ 000000000e307882 facebook::velox::exec::Driver::run(std::shared_ptr<facebook::ve                         lox::exec::Driver>)
                       /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/Driver.cpp:                         768
    @ 000000000e3023fe facebook::velox::exec::Driver::enqueue(std::shared_ptr<facebook                         ::velox::exec::Driver>)::{lambda()#1}::operator()() const
                       /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/Driver.cpp:                         244
    @ 000000000e30cb21 void folly::detail::function::call_<facebook::velox::exec::Driv                         er::enqueue(std::shared_ptr<facebook::velox::exec::Driver>)::{lambda()#1}, true, false                         , void>(, folly::detail::function::Data&)
                       /usr/local/include/folly/Function.h:341
                       -> /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/Driver.c                         pp
    @ 0000000013a6f85a folly::ThreadPoolExecutor::runTask(std::shared_ptr<folly::Threa                         dPoolExecutor::Thread> const&, folly::ThreadPoolExecutor::Task&&)
                       /mnt/DP_disk3/jk/projects/gluten/ep/build-velox/build/velox_ep/                         deps-download/folly/folly/Function.h:370
                       -> /mnt/DP_disk3/jk/projects/gluten/ep/build-velox/build/velox_                         ep/deps-download/folly/folly/executors/ThreadPoolExecutor.cpp
    @ 0000000013a48bb5 folly::CPUThreadPoolExecutor::threadRun(std::shared_ptr<folly::                         ThreadPoolExecutor::Thread>)
                       /mnt/DP_disk3/jk/projects/gluten/ep/build-velox/build/velox_ep/                         deps-download/folly/folly/executors/CPUThreadPoolExecutor.cpp:350
    @ 0000000013a7694a void folly::detail::function::call_<std::_Bind<void (folly::Thr                         eadPoolExecutor::*(folly::ThreadPoolExecutor*, std::shared_ptr<folly::ThreadPoolExecut                         or::Thread>))(std::shared_ptr<folly::ThreadPoolExecutor::Thread>)>, true, false, void>                         (, folly::detail::function::Data&)
                       /usr/include/c++/11/bits/invoke.h:74
                       -> /mnt/DP_disk3/jk/projects/gluten/ep/build-velox/build/velox_                         ep/deps-download/folly/folly/executors/ThreadPoolExecutor.cpp
    @ 000000000bbee078 folly::detail::function::FunctionTraits<void ()>::operator()()
                       /usr/local/include/folly/Function.h:370
                       -> /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/connectors/hi                         ve/FileHandle.cpp
    @ 000000000c4ecf4e folly::NamedThreadFactory::newThread(folly::Function<void ()>&&                         )::{lambda()#1}::operator()()
                       /usr/local/include/folly/executors/thread_factory/NamedThreadFa                         ctory.h:40
                       -> /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/tests/Mu                         ltiFragmentTest.cpp
    @ 000000000c4ffca5 void std::__invoke_impl<void, folly::NamedThreadFactory::newThr                         ead(folly::Function<void ()>&&)::{lambda()#1}>(std::__invoke_other, folly::NamedThread                         Factory::newThread(folly::Function<void ()>&&)::{lambda()#1}&&)
                       /usr/include/c++/11/bits/invoke.h:61
                       -> /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/tests/Mu                         ltiFragmentTest.cpp
    @ 000000000c4ffc60 std::__invoke_result<folly::NamedThreadFactory::newThread(folly                         ::Function<void ()>&&)::{lambda()#1}>::type std::__invoke<folly::NamedThreadFactory::n                         ewThread(folly::Function<void ()>&&)::{lambda()#1}>(std::__invoke_result&&, (folly::Na                         medThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}&&)...)
                       /usr/include/c++/11/bits/invoke.h:96
                       -> /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/tests/Mu                         ltiFragmentTest.cpp
    @ 000000000c4ffc01 void std::thread::_Invoker<std::tuple<folly::NamedThreadFactory                         ::newThread(folly::Function<void ()>&&)::{lambda()#1}> >::_M_invoke<0ul>(std::_Index_t                         uple<0ul>)
                       /usr/include/c++/11/bits/std_thread.h:259
                       -> /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/tests/Mu                         ltiFragmentTest.cpp
    @ 000000000c4ffad5 std::thread::_Invoker<std::tuple<folly::NamedThreadFactory::new                         Thread(folly::Function<void ()>&&)::{lambda()#1}> >::operator()()
                       /usr/include/c++/11/bits/std_thread.h:266
                       -> /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/tests/Mu                         ltiFragmentTest.cpp
    @ 000000000c4ff995 std::thread::_State_impl<std::thread::_Invoker<std::tuple<folly                         ::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}> > >::_M_run                         ()
                       /usr/include/c++/11/bits/std_thread.h:211
                       -> /mnt/DP_disk3/jk/projects/fb-velox/velox/velox/exec/tests/Mu                         ltiFragmentTest.cpp
    @ 00000000000e6792 (unknown)
    @ 0000000000008608 start_thread
    @ 000000000011f352 clone
```

xiaoxmeng commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -41,26 +41,42 @@ RowsStreamingWindowBuild::RowsStreamingWindowBuild(
+}
+
+std::shared_ptr<WindowPartition> RowsStreamingWindowBuild::outputPartition() {
+  return windowPartitions_.front();
```
```
VELOX_CHECK(!windowPartitions_.empty());
return windowPartitions_.front().get();
```

xiaoxmeng commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -102,12 +118,16 @@ void RowsStreamingWindowBuild::noMoreInput() {
 bool RowsStreamingWindowBuild::hasNextPartition() {
-  return !windowPartitions_.empty() &&
-      outputPartition_ + 2 <= windowPartitions_.size();
+  // Remove the processed output partition from the queue.
```
```
  // Remove the processed output partition from the queue.
  while (!windowPartitions_.empty() && outputPartition()->complete() && outputPartition()->numRows() == 0) {
    windowPartitions_.pop_front();
  }
  return !windowPartitions_.empty();
```

xiaoxmeng reviewed (commented):
@JkSelf thanks for fix % comments.

xiaoxmeng commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -101,13 +122,18 @@ void RowsStreamingWindowBuild::noMoreInput() {
 std::shared_ptr<WindowPartition> RowsStreamingWindowBuild::nextPartition() {
-  VELOX_CHECK(hasNextPartition());
-  return windowPartitions_[++outputPartition_];
+  VELOX_CHECK(!windowPartitions_.empty());
```
Remove this as outputPartition has the same check? Thanks!

JkSelf replied:
@xiaoxmeng Yes. Removed.

xiaoxmeng reviewed (approved):
@JkSelf LGTM % nit. Thanks!

aditi-pandit commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -101,13 +122,17 @@ void RowsStreamingWindowBuild::noMoreInput() {
-  return !windowPartitions_.empty() &&
-      outputPartition_ + 2 <= windowPartitions_.size();
+  // Remove the processed output partition from the queue.
+  while (!windowPartitions_.empty() && outputPartition()->complete() &&
```
Does this need to be a while loop ? Partitions always have rows on construction.

aditi-pandit commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -101,13 +122,17 @@ void RowsStreamingWindowBuild::noMoreInput() {
 bool RowsStreamingWindowBuild::hasNextPartition() {
-  return !windowPartitions_.empty() &&
-      outputPartition_ + 2 <= windowPartitions_.size();
+  // Remove the processed output partition from the queue.
```
hasNextPartition() should only be a query API. nextPartition() should be updating the windowPartitions_.

JkSelf replied:
@aditi-pandit @xiaoxmeng If we decide to remove the processed window partition in the nextPartition method, an exception may be thrown in certain cases. This is because hasNextPartition might return true, but when we actually retrieve the window partition in nextPartition, it could be empty after the removal. Therefore, we need to update windowPartition_ within the hasNextPartition method.

aditi-pandit commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -43,24 +43,45 @@ RowsStreamingWindowBuild::RowsStreamingWindowBuild(
+std::shared_ptr<WindowPartition> RowsStreamingWindowBuild::inputPartition()
+    const {
+  VELOX_CHECK(!windowPartitions_.empty());
+  VELOX_CHECK(!windowPartitions_.back()->complete());
```
If we happen to call inputPartition() after the current input is marked complete,  it should still return correctly. What we want to avoid is to addInput to a complete partition, which ensureInputPartition() should take care of. So we can remove 
VELOX_CHECK(!windowPartitions_.back()->complete()); I think.

xiaoxmeng replied:
It is better to have that check as inputPartition() is invoked to add input.

aditi-pandit commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -122,21 +123,19 @@ void RowsStreamingWindowBuild::noMoreInput() {
-  // nextPartition method. This is because hasNextPartition might return true,
-  // but when we actually retrieve the window partition in nextPartition, it
-  // could be empty after the removal.
+  outputPartition_++;
```
Should outputPartition_++ be inside the while loop ? I generally feel the while loop is not needed, but if we are writing that way, then its more logical to increment outputPartition_ in while loop.

xiaoxmeng reviewed (commented):
@JkSelf thanks for the update % few more comments.

xiaoxmeng commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -101,13 +109,28 @@ void RowsStreamingWindowBuild::noMoreInput() {
+      windowPartitions_.front()->numRows() == 0) {
+    windowPartitions_.pop_front();
+  }
+
```
nit: add
```
VELOX_CHECK(hasNextPartition());
```

kagamiori commented on velox/exec/RowsStreamingWindowBuild.cpp:
```diff
@@ -101,13 +109,29 @@ void RowsStreamingWindowBuild::noMoreInput() {
-      outputPartition_ + 2 <= windowPartitions_.size();
+  // Checks if there is a window partition that is either incomplete or
+  // completed but has unconsumed rows.
+  for (const auto& windowPartition : windowPartitions_) {
```
nit: Probably not a big deal, but it looks like the current code iterates through windowPartitions_ from the beginning to the end, while new partitions that are likely incomplete are added from the back, so it may make more sense to loop from the back to the beginning instead? (Please correct me if I'm wrong. Thanks!)

JkSelf replied:
@kagamiori Good catch. It makes sense to me. I have updated the code. Thanks for your review .

kagamiori reviewed (approved):
LGTM.
